### PR TITLE
added SQL aggregate functions to use with QDjangoQuerySet

### PR DESCRIPTION
**Pull request by swex**
hi there.
I think its very useful to make it possible to use aggregate  SQL functions on QDjangoQuerySet.
I've added AggregateType enum to QDjangoWhere, probably its not the base place for it, but it requires much less changes.

### CODE REVIEW

**Comment by jlaine** [comment]
That's really cool! I will try to look at the code and provide feedback

File: src/db/QDjangoQuerySet.h [hunk 1]
```diff
@@ -451,6 +452,20 @@ int QDjangoQuerySet<T>::count() const
     return query.value(0).toInt();
 }
 
+/** Counts the objects in the queryset using an SQL [AVG, COUNT, SUM, MIN, MAX] query,
+ *  or -1 if the query failed.
+ *
+ */
+template <class T>
+QVariant QDjangoQuerySet<T>::aggregate(const QString& field, QDjangoWhere::AggregateType func) const
```
Comment (jlaine) [inline src/db/QDjangoQuerySet.h]: It would seem more intuitive to me to have the aggregation type first, no?

File: src/db/QDjangoQuerySet.h [hunk 2]
```diff
@@ -451,6 +452,20 @@ int QDjangoQuerySet<T>::count() const
...
+{
+    // execute aggregate query
+    QDjangoQuery query(d->aggregateQuery(field,func));
+    if (!query.exec() || !query.next())
+        return -1;
```
Comment (jlaine) [inline src/db/QDjangoQuerySet.h]: Not sure about using -1 as the error, there are legitimate results which will also be -1. Any other ideas?

File: src/db/QDjangoQuerySet_p.h [hunk 3]
```diff
@@ -95,11 +95,15 @@ class QDJANGO_DB_EXPORT QDjangoQuerySetPrivate
 
     // SQL queries
     QDjangoQuery countQuery() const;
+    QDjangoQuery aggregateQuery(const QString &field, QDjangoWhere::AggregateType func) const;
```
Comment (jlaine) [inline src/db/QDjangoQuerySet_p.h]: Let's keep things sorted alphabetically please

File: src/db/QDjangoQuerySet.cpp [hunk 4]
```diff
@@ -343,6 +343,17 @@ bool QDjangoQuerySetPrivate::sqlLoad(QObject *model, int index)
  */
 QDjangoQuery QDjangoQuerySetPrivate::countQuery() const
 {
+    return aggregateQuery("*",QDjangoWhere::COUNT);
+}
+
+/**
+ * @brief QDjangoQuerySetPrivate::aggregateQuery Returns the SQL query to perform a @param field on the current set.
+ * @param field name to aggregae
```
Comment (jlaine) [inline src/db/QDjangoQuerySet.cpp]: typo "aggregae" vs "aggregate"

File: src/db/QDjangoQuerySet.h (hunk 2 above)
Comment (swex) [inline src/db/QDjangoQuerySet.h]: Yeah -1 is stupid, invalid variant would be good I think.

File: src/db/QDjangoQuerySet.h (hunk 1 above)
Comment (swex) [inline src/db/QDjangoQuerySet.h]: yes, you right

File: src/db/QDjangoQuerySet_p.h [hunk 5]
```diff
@@ -95,11 +95,15 @@ class QDJANGO_DB_EXPORT QDjangoQuerySetPrivate
...
     QDjangoQuery deleteQuery() const;
     QDjangoQuery insertQuery(const QVariantMap &fields) const;
     QDjangoQuery selectQuery() const;
     QDjangoQuery updateQuery(const QVariantMap &fields) const;
 
+    //static aggregate functions mapping
+    static QString aggregationToString(QDjangoWhere::AggregateType type);
```
Comment (jlaine) [inline src/db/QDjangoQuerySet_p.h]: I am not convinced we even need a method in the class, a static function in QDjangoQueryset.cpp should be sufficient

File: src/db/QDjangoWhere.h [hunk 6]
```diff
@@ -61,6 +61,13 @@ class QDJANGO_DB_EXPORT QDjangoWhere
         IContains
     };
 
+    enum AggregateType{
+        AVG,
```
Comment (jlaine) [inline src/db/QDjangoWhere.h]: I am undecided as to the right name for these. The uppercase name matches SQL nicely, but doesn't really look very "Qt-ish".

File: src/db/QDjangoQuerySet.cpp [hunk 7]
```diff
@@ -343,6 +343,17 @@ bool QDjangoQuerySetPrivate::sqlLoad(QObject *model, int index)
  */
 QDjangoQuery QDjangoQuerySetPrivate::countQuery() const
```
Comment (jlaine) [inline src/db/QDjangoQuerySet.cpp]: if countQuery becomes this one-liner, I am not convinced we need to keep the method. It can probably be inlined into ::count()

File: src/db/QDjangoQuerySet.cpp [hunk 8]
```diff
@@ -352,14 +363,14 @@ QDjangoQuery QDjangoQuerySetPrivate::countQuery() const
...
-    QString sql = QLatin1String("SELECT COUNT(*) FROM ") + compiler.fromSql();
+
+    QString sql = QLatin1String("SELECT ") + aggregationToString(func)+"("+field+") "+"FROM " + compiler.fromSql();
     if (!where.isEmpty())
         sql += QLatin1String(" WHERE ") + where;
     sql += limit;
     QDjangoQuery query(db);
     query.prepare(sql);
     resolvedWhere.bindValues(query);
-
```
Comment (jlaine) [inline src/db/QDjangoQuerySet.cpp]: unrelated whitespace change :)

**Comment by swex** [comment]
Hello there! Fixed your notes, please check it out again.

**Comment by jlaine** [comment]
Thanks a lot, merged!
